### DISCUSSION TITLE
Internal consumption

### DISCUSSION THREAD

**Discussion by a-teece** [TypQxQ/Sigenergy-Local-Modbus#336] [Ideas]
Is it possible to get or calculate the internal consumption of the Sigenergy system itself? I’m trying to identify a 50-200w unknown load in my house. I’ve added sensors to almost all my loads and it’s still eluding me, leading me to believe it’s the Sig system itself. I’ve now read that it might be self consumption, but would like to track the real number so I can isolate it on my energy dashboard.

**Comment by Roving-Ronin** [comment]
Stick this in your HA Developer Tools --> Template section and see what it gets.  

These lines are just to give you the current state of the sensors, so you can see them and do the maths in your head.
```
PV       {{ (states('sensor.sigen_plant_pv_power') | float(0) * 1000) | round(0) }} W
Grid     {{ (states('sensor.sigen_plant_grid_active_power') | float(0) * 1000) | round(0) }} W
Battery  {{ (states('sensor.sigen_plant_battery_power') | float(0) * 1000) | round(0) }} W
Inverter {{ (states('sensor.sigen_plant_plant_active_power') | float(0) * 1000) | round(0) }} W
```

This is the sensor:
```
template:
  - sensor:
      - name: "Plant - Self Consumption"
        unique_id: sigen_plant_self_consumption
        unit_of_measurement: "W"
        state: >
          {% set pv = states('sensor.sigen_plant_pv_power') | float(0) * 1000 %}
          {% set grid = states('sensor.sigen_plant_grid_active_power') | float(0) * 1000 %}
          {% set battery = states('sensor.sigen_plant_battery_power') | float(0) * 1000 %}
          {% set load = states('sensor.sigen_plant_plant_active_power') | float(0) * 1000 %}

          {# IMPORTANT: normalize battery direction #}
          {% set battery = battery * -1 if battery < 0 else battery %}

          {% set system_loss = pv + grid + battery - load %}

          {{ [system_loss, 0] | max | round(0) }}
```

Instead of adding to a yaml file, I'd add everything after ' state: >' into to a Helper --> Template --> Sensor and assign it to the Sigen Plant. 

<img width="571" height="1140" alt="image" src="https://github.com/user-attachments/assets/7d7e8764-2657-48c0-864f-cd4734f62062" />

That way it will appear as a sensor in the Plant device screen, with all the other official sensors.

<img width="341" height="239" alt="image" src="https://github.com/user-attachments/assets/62cbbbbc-ed5f-4c0f-802f-4ce6a1f00963" />

**Comment by Roving-Ronin** [comment]
If you want to track its kWh usage per day, then use that sensor as the source for an Integral (Riemann) sensor. coverting to kWh.
Just don't use my name typo though.

<img width="575" height="1037" alt="image" src="https://github.com/user-attachments/assets/2ec2d7bc-ab81-4192-ac43-3e5fece3815f" />

Then use the above kWh sensor as the source for a final Helper, being a Utility Meter. Now you've got sensors to see the Active Power and Daily kWh self consumed.

<img width="574" height="1055" alt="image" src="https://github.com/user-attachments/assets/67909467-a8f2-4be7-a1c8-3bb7b4edbfff" />

**Comment by a-teece** [comment]
Thanks for all the tips/help. I’ll grab the laptop and give it a try later today.

**Reply by Roving-Ronin** [reply]
Sorry just looking and it has an error, try this:

```
{% set pv = states('sensor.sigen_plant_pv_power') | float(0) * 1000 %}
          {% set grid = states('sensor.sigen_plant_grid_active_power') | float(0) * 1000 %}
          {% set battery = states('sensor.sigen_plant_battery_power') | float(0) * 1000 %}
          {% set load = states('sensor.sigen_plant_plant_active_power') | float(0) * 1000 %}
          {#
            Energy balance:
            PV + Grid_import - Grid_export + Batt_discharge - Batt_charge = Load + Losses
            Losses = PV + grid - battery - load
            (grid +ve=import, battery +ve=charging)
          #}
          {% set self_consumption = pv + grid - battery - load %}
          {{ [self_consumption, 0] | max | round(0) }}
```

**Reply by a-teece** [reply]
Sorry for the silence. I've been trying to get this going and I'm not sure it's having the desired effect.

Now when the Sig batteries charge overnight, I see a -ve "Untracked consumption" in HA.

<img width="375" height="355" alt="image" src="https://github.com/user-attachments/assets/52316d14-4313-485b-b0d1-892faf2a2f5a" />

**Comment by a-teece** [comment]
I found an earlier discussion on the same topic - poor form not looking there first SORRY - https://github.com/TypQxQ/Sigenergy-Local-Modbus/discussions/201

Anyway, I think i've got the correct data now based on that. Your guide above was still super useful because I haddn't created any intgral or utility meter sensors.

For anyone else looking, the final template I used was:
```
{#
  Taken from the suggestions at https://github.com/TypQxQ/Sigenergy-Local-Modbus/discussions/201
#}


{% set pv1_power = states('sensor.sigen_inverter_pv1_power') | float(0) %}
{% set pv2_power = states('sensor.sigen_inverter_pv2_power') | float(0) %}
{% set pv3_power = states('sensor.sigen_inverter_pv3_power') | float(0) %}

{#
  For solar power there are 2 ways:
   Add up the DC powers per PV string (A)
   Use sensor.sigen_inverter_pv_power or sensor.sigen_plant_pv_power (B)

  Until a recent Sigen inverter firmware update end of july; sigen_inverter_pv_power was also the giving the sum of the DC power per string- so same as (A) Afterwards sigen_inverter_pv_power got the same valueas sigen_plant_pv_power.

  When there is solar, the difference between (A) and (B) are the sum of DC/AC losses + the system own consumption.

  When there is no solar, the own consumption is part of the battery power.
#}

{% set total_pv_power = ((pv1_power + pv2_power)) %}

{% set act_power = states('sensor.sigen_plant_plant_active_power') | float %}
{% set bat_power = states('sensor.sigen_plant_battery_power') | float %}

{#
  Calculate the system power (loss) in kW as follows using the sensor above + active power + battery power
#}

{{ (total_pv_power - bat_power - act_power) | round(3) }}
```
This version works in kW because the other Sigenergy power sensors worked in kW and I wanted to be consistent. I also called it a "Power" sensor for consistency.

<img width="530" height="877" alt="image" src="https://github.com/user-attachments/assets/ce496af2-c598-4441-926f-7a36842aa6a8" />


I then followed your guide to create the Integral Sensor (note that I left the "Metric prefix" blank because I'm already work in kW):
<img width="531" height="943" alt="image" src="https://github.com/user-attachments/assets/08e7e892-76c1-406d-b35f-d0bab8d2ec1d" />

Finally I created the Utility Meter exactly as you described - I called it "Daily Self Consumption Energy" to be consistent with the other sensors included in the integration.

I now have little to zero untracked usage. There is still some around the house, but it is the tiny loads like USB chargers, Alexas, etc.
<img width="1649" height="313" alt="image" src="https://github.com/user-attachments/assets/4798bc74-6506-4e40-9793-ffae786042d4" />
